attestation-service/attestation-service/src/bin/restful-as.rs

```diff
@@ -49,17 +50,30 @@ enum WebApi {
     Policy,
 }
 
+#[derive(Error, Debug)]
+pub enum RestfulError {
+    #[error("Creating service failed")]
+    Service(#[from] ServiceError),
+    #[error("Failed to read AS config file")]
+    Config(#[from] ConfigError),
+    #[error("Openssl errorstack")]
+    Openssl(#[from] openssl::error::ErrorStack),
+    #[error("io error")]
```

Comment:
we'd better add some more details about the error, or include the original error inside the error information, s.t.
```
#[error("io error: {0}")]
```

Reply:
Oh. I use `#[from]` macro like the following
```rust
use thiserror::Error;

#[derive(Error, Debug)]
enum TestError {
    #[error("io error")]
    IO(#[from] std::io::Error)
}


fn t() -> Result<(), TestError> {
    let _a = std::fs::File::open("./nothisfile")?;
    Ok(())
}

fn main() -> Result<(), TestError> {
    let a = t();
    println!("{}", a.expect_err("ss"));
    Ok(())
}
```

The output is 
```
io error
```
without any context.

If convert `#[error("io error")]` to `#[error("io error: {0}")]`, the output will be
```
io error: No such file or directory (os error 2)
```

Reply:
oh, that's bad. so, it's up to the logger impl to print the cause or not. We could add  `{0}` to the message. afaiu this will just call the .fmt() of the wrapped struct, in this case the error cause.

I found it a bit unfortunate to see those deeply nested log lines e.g. in API responses ("Error: ... Error: ... Error ...") that look like flattened backtraces and maybe even reveal too much of the internals on an API endpoint. They are sometimes called "Big Ball of Mud" errors: (see [this chapter](https://www.lpalmieri.com/posts/error-handling-rust/#avoid-ball-of-mud-error-enums) from the the zero2prod rust book). They are bad because:

> We do not want to expose the implementation details of the fallible routines that get mapped to UnexpectedError by subscribe - it must be opaque.

E.g. you don't want to log the rate limit error that you get from your managed database backend to the consumer.

There seems to be consensus that duplicating the source on the error message is an antipattern, see this [SO answer](https://stackoverflow.com/questions/62869360/should-an-error-with-a-source-include-that-source-in-the-display-output).

That said, if the logger doesn't respect "source", I don't know how this information can still be surfaced in the log output, but limited to the consumer, especially with the autogenerated RPC code. I'll open a dedicated issue about this, and it should be fine in the PR.

Reply:
Overall, I do not like the flattened format. But compared to a non-detailed error information, I have to choose the flatten one.

You are right that the error information should care about the boundary, e.g. `AttestationService` lib should response a detailed error information, while gRPC/RESTful responsed error should not expose internal details.

To me, if way 2 in https://stackoverflow.com/a/63620301 would be better in crates (e.g. "failed to open file XXX" is better than "io error", which will do great help to debug), while we can cut the details of the errors in concrete binary implementations to avoid explosure of internal information.

Reply:
apparently, the log create doesn't support structured logging yet (looks like they plan to add it, though). fwiw in the the `tracing` crate it looks like this:

```rust
        error!(error = %e, source = e.source(), "could not do this");
``` 

which will produce:

```bash
2024-03-04T10:41:17.556335Z ERROR rust_error_logging: could not do this error=some thing with context source=some other thing failed
```

Reply:
Well. This could be a problem but up to now it is ok to me without structured information.